stan-dev/stan
# Bugfix/issue 2146 compound declare define matrix

#### Submisison Checklist


#### Summary

Fixed Stan lang grammar rules so that compound declaration/definition statements for matrix variables record variable definition during parsing.

#### Intended Effect

Compount declaration/definition statements work for matrix types.

#### How to Verify

Run the Stan compiler over test model src/test/test-models/good/declare-define-var-matrix-2.stan, which contains the following 2 matrix definition statements:

> transformed data {
>   matrix[2,3] m23;
>   matrix[2,3] yam23 = m23;
> }

Verify that the compiled model includes variable assignment statement as well as variable declaration:

>         // initialize undefined transformed variables to avoid seg fault on val access
>         stan::math::fill(m23,DUMMY_VAR__);
>         stan::math::fill(yam23,DUMMY_VAR__);
>         // assign variable definitions
>         stan::math::assign(yam23,m23);

#### Side Effects

#### Documentation

#### Reviewer Suggestions

Daniel Lee

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Mitzi Morris

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)
